File: Lista01/PY/ex12.py
class Produto:
    def __init__(self, nome: str, preco: float):
        self.nome = nome
        self.preco = preco

    def __add__(self, outro):
        """Sobrecarga do operador + para somar os preços de dois produtos"""
        if isinstance(outro, Produto):
            return Produto("Combinação", self.preco + outro.preco)
        return NotImplemented

    def __str__(self):
        return f"{self.nome}: R$ {self.preco:.2f}"

# Exemplo de uso
produto1 = Produto("Produto A", 10.0)
produto2 = Produto("Produto B", 20.0)
produto_comb = produto1 + produto2

print(produto1)         # Saída: Produto A: R$ 10.00
print(produto2)         # Saída: Produto B: R$ 20.00
print(produto_comb)     # Saída: Combinação: R$ 30.00

File: Lista01/Java/ex12.java
public class ex12 {
    // Classe Produto
    private String nome;
    private double preco;

    public ex12(String nome, double preco) {
        this.nome = nome;
        this.preco = preco;
    }

    public ex12 somar(ex12 outro) {
        return new ex12(this.nome + " & " + outro.nome, this.preco + outro.preco);
    }

    @Override
    public String toString() {
        return "Produto: " + nome + ", Preço: R$" + String.format("%.2f", preco);
    }

    public static void main(String[] args) {
        ex12 produto1 = new ex12("Produto A", 50.0);
        ex12 produto2 = new ex12("Produto B", 30.0);

        ex12 produtoCombinado = produto1.somar(produto2);
        System.out.println(produtoCombinado); // Produto: Produto A & Produto B, Preço: R$80.00
    }
}
